compute-nodes.yaml

```diff
@@ -0,0 +1,94 @@
+- import_playbook: common.yaml
+
+- hosts: all
+  gather_facts: no
+
+  tasks:
+  # Build a list of worker nodes with indexes
+  - name: Build worker list
+    set_fact:       
+      worker_list: "{{ worker_list | default([]) + [ item.1 + '-' + item.0 | string ] }}"
+    with_indexed_items: "{{ [os_port_worker] * os_compute_nodes_number }}"
+
+  # Create the SDN/primary port for each worker node
+  - name: 'Create the Compute ports'
+    os_port:
+      name: "{{ item }}"
+      network: "{{ os_network }}"
+      security_groups:
+      - "{{ os_sg_worker }}"
+      allowed_address_pairs:
+      - ip_address: "{{ os_ingressVIP }}"
+    with_items: "{{ worker_list }}"
+    register: ports
+
+  # Tag each SDN/primary port with cluster name
+  - name: 'Set Compute ports tag'
+    command:
+      cmd: "openstack port set --tag {{ cluster_id_tag }} {{ item }}"
+    with_items: "{{ worker_list }}"  
+
+  # Ensure that each network specified in additionalNetworks exists
+  - name: Verify additionalNetworks
+    os_networks_info:
+      name: "{{ item.id }}"
+    with_items: "{{ additionalNetworks }}"
+    register: result
+
+  # Expand additionalNetworks by the count parameter in each network definition
+  - name: Build port and port index list for additionalNetworks
+    set_fact:
+      port_list: "{{ port_list | default([]) + [ { 
+                      'net_name' : item.1.id,
+                      'uuid' : result.results[item.0].openstack_networks[0].id, 
+                      'type' : item.1.type|default('normal')
+                      } ] * item.1.count|default(1) }}"
+      index_list: "{{ index_list | default([]) + range(item.1.count|default(1)) | list }}"
+    with_indexed_items: "{{ additionalNetworks }}"
+
+  # Calculate and save the name of the port
+  # The format of the name is cluster_name-worker-workerID-networkUUID(partial)-count
+  # i.e. fdp-nz995-worker-1-99bcd111-1
+  - name: Save port name
+    set_fact:
+      port_name_list: "{{ port_name_list | default([]) + [ item.1 | combine( {'name' : item.1.uuid | regex_search('([^-]+)') + '-' + index_list[item.0]|string } ) ] }}" 
+    with_indexed_items: "{{ port_list }}"
```

Comment:
This is fine as is and probably working.
However it would be easier to read and probably maintain if we don't over-use `set_fact` tasks that use a lot of filters. They are expensive at scale (memory + CPU) and we usually replace them with a module (pure Python) or an Action plugin.

I would be happy to help doing that, but I wanted to share this feedback here. Not blocking it but would be nice to have at some point.

Reply:
Thanks for the feedback!  I am quite sure that a lot of my pretzel code could be simplified :).  I found it quite hard to do something that I view as simple, make a nested loop count dependent upon a value in the data you are looping over.  I.e., I have count: in the additionalNetworks variable and I could not find a good way to base inner loops on count:.  While I have done a fair amount of automation with Ansible I have never used custom python modules.  As a future optimization, I would be happy to optimize.